hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/SplitTableRegionProcedure.java

```diff
@@ -110,6 +111,15 @@ public SplitTableRegionProcedure(final MasterProcedureEnv env,
     // we fail-fast on construction. There it skips the split with just a warning.
     checkOnline(env, regionToSplit);
     this.bestSplitRow = splitRow;
+    TableDescriptor tableDescriptor = env.getMasterServices().getTableDescriptors()
+      .get(getTableName());
+    Configuration conf = env.getMasterConfiguration();
+    if (hasBestSplitRow()) {
```

Comment:
We do not have this logic in the past? What is changed so now we need to apply this restriction in SplitTableRegionProcedure? IIRC the logic is done at region server side?

Reply:
Yes, we didn't have this logic in the past. I think we can apply the restriction to a user-specified split point because without this logic, we can easily break the restriction by splitting with specifying a split point. And the user-specified split point is passed to the Master side, we need to do it on the master side.

What do you think? @Apache9

Reply:
Finally we should get the actual split point back from region server? No? Then this should be a bug for the current code base?

Reply:
> Finally we should get the actual split point back from region server? No? 

No, I don't think so.

Let's see we have a table that has a key prefix restriction where the prefix length is 2 bytes.
When a user runs split command with specifying a split point `abc` in the hbase shell, this will break the key prefix restriction if we split the region by `abc`. So I think we can apply the restriction to the user-specified split point, and the restriction-applied split point will be `ab`, which won't break the restriction.

Reply:
@Apache9 What about this? Thanks.

Reply:
But what if we do not provide a split point when constructing the SplitTableRegionProcedure? How do we get the split point?

Reply:
In that case, `hasBestSplitRow()` returns `false` and SplitTableRegionProcedure will try to find a best split key from RS as follows:
https://github.com/apache/hbase/blob/781da1899aad7a50ce9c096be757cc1cab3dc15e/hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/SplitTableRegionProcedure.java#L193-L199

In this change, we apply a split restriction only when a user specifies a split point (and `hasBestSplitRow()` returns `true`).

Reply:
OK, so this should be a behavior change. In the past if user specify a split point, we will always use it no matter whether it breaks the split point restriction and now we will try to convert it. Is the old behavior a bug or a feature? For me I do not think we should let user to break the restriction. @saintstack WDYT sir?

Reply:
Good point. So user may be 'surprised' if we do not split where they want? Will there be a message saying so anywhere that their choice has been over-ruled by the restriction? Or will it be obvious that the 'restriction' over-ruled?

I'm good w/ the restriction over-ruling the user as long as there a log to this effect (add the 'behavior change' to the existing nice release note @brfrn169 )

Reply:
I think we can add a WARN message in the Master log when the user-specified split point is over-ruled by the restriction. I will do that. And I will add the 'behavior change' to the release note. Thanks.

Reply:
So here we will only use SplitRestriction to fix the split row? Then what if users uses the deprecated KeyPrefixSplitPolicy? We will not fix the split row if it breaks the rule?

Reply:
Yes, that's the behavior of KeyPrefixSplitPolicy, and we will not fix the split row even if it breaks the rule. And maybe it's not easy to fix it because RegionSplitPolicy doesn't have any method to restrict/convert a user-specified split point. It has only `byte[] getSplitPoint()` that gets an appropriate split point calculated based on its policy.

Reply:
OK, so in fact we are not changing the behavior? If you use the old KeyPrefixSplitPolicy, there is nothing changed. If you use the new SplitRestriction, then you will find out that you are not allowed to break the restriction when proposing a split point. Could mention this in the release note.